### PR TITLE
Do not error when thumbnailing invalid files

### LINKED ISSUE #3763: Errors from thumbnailing due to bad format give a 500
We should either accept it regardless (with no thumbnail) or give a 400. Example:

```
2018-08-28 09:36:53,588 - synapse.http.server - 101 - ERROR - POST-10386 - Failed handle request via <function _async_render_POST at 0x7f0426db1cf8>: <XForwardedForRequest at 0x7f03ecea0440 method='POST' uri='/_matrix/media/v1/upload?filename=1535448124087775917601.jpg' clientproto='HTTP/1.1' site=8085>: Traceback (
most recent call last):
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 654, in _runCallbacks
    current.result = callback(current.result, *args, **kw)
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1475, in gotResult
    _inlineCallbacks(r, g, status)
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1416, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/python/failure.py", line 491, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
--- <exception caught here> ---
  File "synapse/http/server.py", line 81, in wrapped_request_handler
    
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1416, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/python/failure.py", line 491, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "synapse/rest/media/v1/upload_resource.py", line 95, in _async_render_POST
    
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1416, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/python/failure.py", line 491, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "synapse/rest/media/v1/media_repository.py", line 174, in create_content
    
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1416, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/python/failure.py", line 491, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "synapse/rest/media/v1/media_repository.py", line 626, in _generate_thumbnails
    
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/python/threadpool.py", line 250, in inContext
    result = inContext.theWork()
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/python/threadpool.py", line 266, in <lambda>
    inContext.theWork = lambda: context.call(ctx, func, *args, **kw)
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/python/context.py", line 122, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/python/context.py", line 85, in callWithContext
    return func(*args,**kw)
  File "synapse/rest/media/v1/thumbnailer.py", line 60, in scale
    
  File "synapse/rest/media/v1/thumbnailer.py", line 96, in _encode_image
    
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/PIL/Image.py", line 1950, in save
    save_handler(self, fp, filename)
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/PIL/JpegImagePlugin.py", line 623, in _save
    raise IOError("cannot write mode %s as JPEG" % im.mode)
exceptions.IOError: cannot write mode RGBA as JPEG
```

### LINKED ISSUE #4379: builtins.OSError: cannot identify image file '/var/lib/synapse/media_store/local_content/[…]'
### Description

Synapse log are spitting errors related to two images in `local_content`. They are both SVG.

Full excerpt:
```
synapse.http.server: [GET-9] Failed handle request via <function ThumbnailResource._async_render_GET at 0x6ecff9720d90>: <XForwardedForRequest at 0x6ecff5202320 method='GET' uri='/_matrix/media/v1/thumbnail/activis.me/TducGElbUrnYortubeohKFae?width=48&height=48&method=crop' clientproto='HTTP/1.1' site=8008>: Traceback (most recent call last):
  File "/usr/lib/python3.7/site-packages/twisted/internet/defer.py", line 1418, in _inlineCallbacks
    result = g.send(result)
  File "/usr/lib/python3.7/site-packages/synapse/rest/media/v1/thumbnail_resource.py", line 171, in _select_or_generate_local_thumbnail
    url_cache=media_info["url_cache"],
  File "/usr/lib/python3.7/site-packages/twisted/internet/defer.py", line 1613, in unwindGenerator
    return _cancellableInlineCallbacks(gen)
  File "/usr/lib/python3.7/site-packages/twisted/internet/defer.py", line 1529, in _cancellableInlineCallbacks
    _inlineCallbacks(None, g, status)
--- <exception caught here> ---
  File "/usr/lib/python3.7/site-packages/synapse/http/server.py", line 81, in wrapped_request_handler
    yield h(self, request)
  File "/usr/lib/python3.7/site-packages/synapse/rest/media/v1/thumbnail_resource.py", line 67, in _async_render_GET
    request, media_id, width, height, method, m_type
  File "/usr/lib/python3.7/site-packages/synapse/rest/media/v1/thumbnail_resource.py", line 171, in _select_or_generate_local_thumbnail
    url_cache=media_info["url_cache"],
  File "/usr/lib/python3.7/site-packages/twisted/internet/defer.py", line 1418, in _inlineCallbacks
    result = g.send(result)
  File "/usr/lib/python3.7/site-packages/synapse/rest/media/v1/media_repository.py", line 466, in generate_local_exact_thumbnail
    thumbnailer = Thumbnailer(input_path)
  File "/usr/lib/python3.7/site-packages/synapse/rest/media/v1/thumbnailer.py", line 32, in __init__
    self.image = Image.open(input_path)
  File "/usr/lib/python3.7/site-packages/PIL/Image.py", line 2657, in open
    % (filename if filename else fp))
builtins.OSError: cannot identify image file '/var/lib/synapse/media_store/local_content/Td/uc/GElbUrnYortubeohKFae'
synapse.http.server: [GET-10] Failed handle request via <function ThumbnailResource._async_render_GET at 0x6ecff9720d90>: <XForwardedForRequest at 0x6ecff5202828 method='GET' uri='/_matrix/media/v1/thumbnail/activis.me/VryPzhdsbZOaAjkFUhTmRfqr?width=48&height=48&method=crop' clientproto='HTTP/1.1' site=8008>: Traceback (most recent call last):
  File "/usr/lib/python3.7/site-packages/twisted/internet/defer.py", line 1418, in _inlineCallbacks
    result = g.send(result)
  File "/usr/lib/python3.7/site-packages/synapse/rest/media/v1/thumbnail_resource.py", line 171, in _select_or_generate_local_thumbnail
    url_cache=media_info["url_cache"],
  File "/usr/lib/python3.7/site-packages/twisted/internet/defer.py", line 1613, in unwindGenerator
    return _cancellableInlineCallbacks(gen)
  File "/usr/lib/python3.7/site-packages/twisted/internet/defer.py", line 1529, in _cancellableInlineCallbacks
    _inlineCallbacks(None, g, status)
--- <exception caught here> ---
  File "/usr/lib/python3.7/site-packages/synapse/http/server.py", line 81, in wrapped_request_handler
    yield h(self, request)
  File "/usr/lib/python3.7/site-packages/synapse/rest/media/v1/thumbnail_resource.py", line 67, in _async_render_GET
    request, media_id, width, height, method, m_type
  File "/usr/lib/python3.7/site-packages/synapse/rest/media/v1/thumbnail_resource.py", line 171, in _select_or_generate_local_thumbnail
    url_cache=media_info["url_cache"],
  File "/usr/lib/python3.7/site-packages/twisted/internet/defer.py", line 1418, in _inlineCallbacks
    result = g.send(result)
  File "/usr/lib/python3.7/site-packages/synapse/rest/media/v1/media_repository.py", line 466, in generate_local_exact_thumbnail
    thumbnailer = Thumbnailer(input_path)
  File "/usr/lib/python3.7/site-packages/synapse/rest/media/v1/thumbnailer.py", line 32, in __init__
    self.image = Image.open(input_path)
  File "/usr/lib/python3.7/site-packages/PIL/Image.py", line 2657, in open
    % (filename if filename else fp))
builtins.OSError: cannot identify image file '/var/lib/synapse/media_store/local_content/Vr/yP/zhdsbZOaAjkFUhTmRfqr'
```

### Version information

- **Homeserver**: mine
- **Version**: 0.34.1.1
- **Install method**: Package manager
- **Platform**: ArchLinux

### PR DESCRIPTION
**Pull request by clokep**
This attempts to fix a Sentry error we've been seeing a lot: https://sentry.matrix.org/sentry/synapse-matrixorg/issues/120394/?query=is%3Aunresolved

Pretty much if Pillow fails to load a file we raise an exception and 500. This swallows the exception and returns a 404 (saying we cannot thumbnail the image).

I believe that this fixes #3763 and fixes #4379.

### CODE REVIEW

File: tests/rest/media/v1/test_media_storage.py [hunk 1]
```diff
@@ -325,11 +336,15 @@ def _test_thumbnail(self, method, expected_body):
...
+                    channel.result["body"], expected_body, channel.result["body"]
+                )
+            else:
+                # ensure that the result is at least some valid image
+                Image.open(BytesIO(channel.result["body"]))
         else:
-            # ensure that the result is at least some valid image
-            Image.open(BytesIO(channel.result["body"]))
+            # A 404 with no body.
+            self.assertEqual(channel.code, 404)
```
Comment (clokep) [inline tests/rest/media/v1/test_media_storage.py:350]: This actually *does* return a body weird:

```json
{"errcode":"M_NOT_FOUND","error":"Not f[65 chars]']"}
```

**Review by turt2live — CHANGES_REQUESTED** [review]
400 error please, for spec compliance :)

https://matrix.org/docs/spec/client_server/r0.6.1#get-matrix-media-r0-thumbnail-servername-mediaid

> Status code 400:
> The request does not make sense to the server, or the server cannot thumbnail the content. For example, the client requested non-integer dimensions or asked for negatively-sized images.

**Comment by clokep** [comment]
To be fair, this returns the same error that Synapse *has been returning* in these situations. It just does error handling better.

**Comment by clokep** [comment]
Looking more closely it seems these endpoints should never return a 404...

**Comment by turt2live** [comment]
sounds like a good time to bring it into compliance then 😇

**Comment by turt2live** [comment]
(they can return a 404 if it doesn't know about the media ID/server combination, or can't get access to the media from a remote server)

**Comment by clokep** [comment]
> (they can return a 404 if it doesn't know about the media ID/server combination, or can't get access to the media from a remote server)

It does not look like any of the media endpoints mention this in the spec.

**Comment by clokep** [comment]
I believe I've made the additional changes to return a 400 error in this case. I can split that out to a separate PR if desired.

**Comment by turt2live** [comment]
> It does not look like any of the media endpoints mention this in the spec.

Indeed, it's one of the ~~few~~ many places we can improve.

File: synapse/rest/media/v1/_base.py [hunk 2]
```diff
@@ -69,13 +70,30 @@ def parse_media_id(request):
...
-        request,
-        404,
-        cs_error("Not found %r" % (request.postpath,), code=Codes.NOT_FOUND),
-        send_cors=True,
-    )
+def respond_404(request: Request):
+    respond_with_code(request, 404)
+
+
+def respond_with_code(request: Request, code: int):
```
Comment (turt2live) [inline synapse/rest/media/v1/_base.py:77]: fwiw, the thumbnail API returning JSON for errors is expected under the Standard Error Response.

**Review by turt2live — APPROVED** [review]
lgtm spec-wise, assuming it does return the Standard Error Response

File: synapse/rest/media/v1/_base.py (hunk 2 above)
Comment (clokep) [reply synapse/rest/media/v1/_base.py:77]: That really doesn't makes semantic sense for an API, but if that's what the spec says I can revert this change.

**Comment by clokep** [comment]
The tests here seem to only test the remote case, which I'm not a big fan of. I'll look to expand those for local thumbnails too.

File: tests/rest/media/v1/test_media_storage.py (hunk 1 above)
Comment (clokep) [reply tests/rest/media/v1/test_media_storage.py:346]: This is on purpose, I've updated the test (and this comment).

File: synapse/rest/media/v1/media_repository.py [hunk 3]
```diff
@@ -506,14 +522,35 @@ async def generate_local_exact_thumbnail(
...
+        try:
+            thumbnailer = Thumbnailer(input_path)
+        except ThumbnailError:
+            logger.warning(
+                "Unable to generate a thumbnail for remote media %s from %s using a method of %s and type of %s",
+                media_id,
+                server_name,
+                t_method,
+                t_type,
+            )
```
Comment (erikjohnston) [inline synapse/rest/media/v1/media_repository.py:551]: Do we want log why as well?

**Review by erikjohnston — APPROVED** [review]

File: synapse/rest/media/v1/media_repository.py (hunk 3 above)
Comment (clokep) [reply synapse/rest/media/v1/media_repository.py:551]: Hmm....that's a good thought, we'd have to pass the original exception through (I guess using `raise from`, maybe?)

Comment (clokep) [reply synapse/rest/media/v1/media_repository.py:553]: @erikjohnston I'm hoping b5b96ea65731f3eb86d5c06b7defd1ab3c23d004 is what you were thinking?

**Pull request merged**